### ISSUE TITLE
Kitty's new keyboard protocol support

### ISSUE THREAD

**Issue by zeertzjq** [neovim/neovim#14400]
Kitty recently introduced a new keyboard protocol that supports representing many more keys and key combinations. (<https://sw.kovidgoyal.net/kitty/keyboard-protocol/>) Adding support for this protocol may be helpful in some circumstances.

**Comment by zeertzjq** [comment]
Probably related: #2204

**Comment by justinmk** [comment]
related: https://github.com/neovim/neovim/issues/5916

**Comment by zeertzjq** [comment]
There are also two issues regarding `modifyOtherKeys` (#14350 and #15352) that may be related to this issue.

**Comment by rumpelsepp** [comment]
foot started implementing this protocol: https://codeberg.org/dnkl/foot/pulls/811

**Comment by neovim-discourse** [comment]
This issue has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/nitty-gritty-questions-on-terminal-nvim-and-keyboard-events/1624/1

**Comment by tmccombs** [comment]
With the improvements around distinguishing ambiguous keys in the 0.7 release, how much would need to be done to support the first level of the kitty keyboard protocol ("disambiguate escape codes")? Would it just be sending the correct CSI codes to the terminal (`CSI > 1u` and `CSI < u` to push and pop state, and `CSI ? u` to query support)? Or would it be more involved?

**Comment by zeertzjq** [comment]
It is a bit tricky to encode new modifiers, especially if considering backward-compatibility.

**Comment by tmccombs** [comment]
Could neovim just ignore modifiers it doesn't recognize?

**Comment by kovidgoyal** [comment]
Yes of course, how it converts the key events from the terminal to its own internal key
event representation is entirely up to it.

**Comment by benfrain** [comment]
Would be great to see support for say, <kbd>Hyper</kbd> in Neovim with Kitty.

**Comment by gpanders** [comment]
> With the improvements around distinguishing ambiguous keys in the 0.7 release, how much would need to be done to support the first level of the kitty keyboard protocol ("disambiguate escape codes")? Would it just be sending the correct CSI codes to the terminal (`CSI > 1u` and `CSI < u` to push and pop state, and `CSI ? u` to query support)? Or would it be more involved?

https://github.com/neovim/neovim/pull/18181

**Comment by tmccombs** [comment]
Thank you @gpanders !

**Comment by zeertzjq** [comment]
This isn't fully supported yet.

**Comment by ModProg** [comment]
One thing I noticed, not sure where to put it, that the differentiation between `<Tab>` and `<C-i>` only exists in normal mode and not in insert mode.

I tried:
```vim
imap <TAB> hello
imap <C-i> test
```
Only one of them is in effect for both key presses.

But in normal mode
```vim
nmap <TAB> dd
nmap <C-i> yy
```
works as expected.

**Comment by zeertzjq** [comment]
> One thing I noticed, not sure where to put it, that the differentiation between `<Tab>` and `<C-i>` only exists in normal mode and not in insert mode.
> 
> I tried:
> 
> ```viml
> imap <TAB> hello
> imap <C-i> test
> ```
> 
> Only one of them is in effect for both key presses.

This shouldn't happen.

**Comment by ModProg** [comment]
I think my report was invalid, I forgot to test with `-u NONE`, with it, it works correctly.

The problem seamed to be my auto complete plugin that was not correctly mapping when not currently completing.
Therefor not a neovim bug, but a bug on their side.

**Comment by hacker-DOM** [comment]
I am not able to get this functionality. I have:

```vim
nnoremap <C-e> :echomsg('unshifted e')<CR>
nnoremap <C-S-e> :echomsg('shifted E')<CR>
nnoremap g<C-e> ge
nnoremap g<C-S-e> gE
```

The distinction is correctly made in the plain (echomsg) case, but not in the chord (g) case. In particular, it works for `g<C-e>`, but nothin happens for `g<C-S-e>`

I am using Kitty 0.26.3, Nvim 0.8.0

Thanks in advance!

**Comment by eugenesvk** [comment]
By the way, would implementing this protocol finally allow you to set left and right modifier keys separately?

**Comment by hacker-DOM** [comment]
It won't afaict; but it should allow you to set shortcuts to bindings like Cmd+C etc, which is not currently natively supported

**Comment by zeertzjq** [comment]
<kbd>Super</kbd> is supported as of #24357. <kbd>Hyper</kbd> will be a bit complicated.

**Comment by JosefLitos** [comment]
> By the way, would implementing this protocol finally allow you to set left and right modifier keys separately?

from what I've tried, the protocol does support it in the form of sending separate presses of the modifiers (for example: `57447` for right and `57441` for left shift) which would need internal state holding from the program side, but it does send it - try `\033[>11u`.

**Comment by justinmk** [comment]
@MartyMcFlyInTheSky the checklist at the top of the issue indicates the state. Ask usage questions on https://github.com/neovim/neovim/discussions , any more comments will be deleted here.

**Comment by wongjiahau** [comment]
This is crucial for emulating Ki's Momentary Layer feature in Neovim
